services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+            url: self.api_url.clone(),
+            response: text,
+            error,
+        })
+    }
+
+    fn time_range_request(time: &TimeInterval) -> Result<(DateTime<Utc>, DateTime<Utc>)> {
+        let t_start =
+            time.start()
+                .as_utc_date_time()
+                .ok_or(geoengine_operators::error::Error::DataType {
+                    source: geoengine_datatypes::error::Error::NoDateTimeValid {
+                        time_instance: time.start(),
+                    },
+                })?;
+
+        // shift start by 1 day to ensure getting the most recent data for start time
+        let t_start = t_start - Duration::days(1);
+
+        let t_end =
+            time.end()
+                .as_utc_date_time()
+                .ok_or(geoengine_operators::error::Error::DataType {
+                    source: geoengine_datatypes::error::Error::NoDateTimeValid {
+                        time_instance: time.end(),
+                    },
+                })?;
+
+        Ok((t_start, t_end))
+    }
+}
+
+#[async_trait]
+impl MetaData<GdalLoadingInfo, RasterResultDescriptor> for SentinelS2L2aCogsMetaData {
+    async fn loading_info(
+        &self,
+        query: QueryRectangle,
+    ) -> geoengine_operators::util::Result<GdalLoadingInfo> {
+        // TODO: propagate error properly
+        self.create_loading_info(query).await.map_err(|e| {
+            geoengine_operators::error::Error::LoadingInfo {
+                reason: format!("{:?}", e),
+            }
+        })
+    }
+
+    async fn result_descriptor(&self) -> geoengine_operators::util::Result<RasterResultDescriptor> {
+        Ok(RasterResultDescriptor {
+            data_type: self.band.data_type,
+            spatial_reference: SpatialReference::new(
+                SpatialReferenceAuthority::Epsg,
+                self.zone.epsg,
+            )
+            .into(),
+            measurement: Measurement::Unitless,
+            no_data_value: self.band.no_data_value,
+        })
+    }
+
+    fn box_clone(&self) -> Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor>> {
+        Box::new(self.clone())
+    }
+}
+
+#[async_trait]
+impl MetaDataProvider<GdalLoadingInfo, RasterResultDescriptor> for SentinelS2L2aCogsDataProvider {
+    async fn meta_data(
+        &self,
+        dataset: &DatasetId,
+    ) -> Result<
+        Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor>>,
+        geoengine_operators::error::Error,
+    > {
+        let dataset = self
+            .datasets
+            .get(&dataset)
+            .ok_or(geoengine_operators::error::Error::UnknownDatasetId)?;
+
+        Ok(Box::new(SentinelS2L2aCogsMetaData {
+            api_url: self.api_url.clone(),
+            zone: dataset.zone.clone(),
+            band: dataset.band.clone(),
+        }))
+    }
+}
+
+#[async_trait]
+impl MetaDataProvider<MockDatasetDataSourceLoadingInfo, VectorResultDescriptor>
```

Comment:
do we really have to implement all kinds if they are not really implemented? There should be a task to solve this

Reply:
a case for trait specialization... maybe we can create a macro that builds NotImplented-stubs

Reply:
i'm still not convinced that we need to require all the implementations everyehere. This is what happens if you do this. i guess there should be a better metadata provider instantiation / retrival mechanisim

Reply:
I can't see how that's possible, you would have to explain me how to

Reply:
yes but that is a another task